src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/Tensor.cs

```diff
@@ -1,751 +1,5624 @@
+        /// </summary>
+        /// <param name="tensors">Input <see cref="Tensor{T}"/>.</param>
+        public static Tensor<T> Stack<T>(params ReadOnlySpan<Tensor<T>> tensors)
+        {
+            return StackAlongDimension(0, tensors);
+        }
+
+        /// <summary>
+        /// Join multiple <see cref="Tensor{T}"/> along a new dimension. The axis parameter specifies the index of the new dimension. All tensors must have the same shape.
+        /// </summary>
+        /// <param name="tensors">Input <see cref="Tensor{T}"/>.</param>
+        /// <param name="dimension">Index of where the new dimension will be.</param>
+        public static Tensor<T> StackAlongDimension<T>(int dimension, params ReadOnlySpan<Tensor<T>> tensors)
+        {
+            if (tensors.Length < 2)
+                ThrowHelper.ThrowArgument_StackTooFewTensors();
+
+            for (int i = 1; i < tensors.Length; i++)
+            {
+                if (!tensors[0].Lengths.SequenceEqual(tensors[i].Lengths))
+                    ThrowHelper.ThrowArgument_StackShapesNotSame();
+            }
+
+            if (dimension < 0)
+                dimension = tensors[0].Rank - dimension;
+
+            Tensor<T>[] outputs = new Tensor<T>[tensors.Length];
+            for (int i = 0; i < tensors.Length; i++)
+            {
+                outputs[i] = Tensor.Unsqueeze(tensors[i], dimension);
+            }
+            return Tensor.ConcatenateOnDimension<T>(dimension, outputs);
+        }
+
+        /// <summary>
+        /// Join multiple <see cref="Tensor{T}"/> along a new dimension that is added at position 0. All tensors must have the same shape.
+        /// </summary>
+        /// <param name="tensors">Input <see cref="Tensor{T}"/>.</param>
+        /// <param name="destination"></param>
+        public static ref readonly TensorSpan<T> Stack<T>(scoped in ReadOnlySpan<Tensor<T>> tensors, in TensorSpan<T> destination)
+        {
+            return ref StackAlongDimension(tensors, destination, 0);
+        }
+
+        /// <summary>
+        /// Join multiple <see cref="Tensor{T}"/> along a new dimension. The axis parameter specifies the index of the new dimension. All tensors must have the same shape.
+        /// </summary>
+        /// <param name="tensors">Input <see cref="Tensor{T}"/>.</param>
+        /// <param name="destination"></param>
+        /// <param name="dimension">Index of where the new dimension will be.</param>
+        public static ref readonly TensorSpan<T> StackAlongDimension<T>(scoped ReadOnlySpan<Tensor<T>> tensors, in TensorSpan<T> destination, int dimension)
+        {
+            if (tensors.Length < 2)
+                ThrowHelper.ThrowArgument_StackTooFewTensors();
+
+            for (int i = 1; i < tensors.Length; i++)
+            {
+                if (!tensors[0].Lengths.SequenceEqual(tensors[i].Lengths))
+                    ThrowHelper.ThrowArgument_StackShapesNotSame();
+            }
+
+            if (dimension < 0)
```

Comment:
ditto

Reply:
it would throw before, but I've added it locally with a better error message for a followup pr.